text/0020-block-proposer-improvements.md

```diff
@@ -0,0 +1,108 @@
+# Block Proposer Improvements
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0020](https://github.com/casperlabs/ceps/pull/0020)
+
+This CEP describes suggested modifications to the Block Proposer component designed to fix some of the known problems with it.
+
+## Motivation
+
+[motivation]: #motivation
+
+There are a few known problems in the Block Proposer related to event ordering. The component is written with the assumption that it will receive events in the same order they were generated, but because of the asynchronous nature of event handling within the reactor, this assumption is sometimes incorrect.
+
+There are two main ways in which a violation of the underlying assumptions can cause issues with the Block Proposer:
+
+- it can see blocks being finalized before it sees them proposed
+- it can be asked to propose a block before it receives a notification about a block being proposed by another validator, leading to duplications of deploys.
+
+A separate solution for each of these two issues will be described below.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+### Seeing finalization before proposal
+
+The solution to the first issue should be fairly straightforward: buffer finalized blocks that haven't been proposed as far as we know, and then handle both proposal and finalization when we see the proposal.
```

Comment:
What if we have e.g. lots of empty blocks? They will all have identical `ProtoBlock`s, and they might all get finalized. Will that cause us to remove the "empty" entry from the proposed blocks, and then buffer it forever if another empty block gets finalized?

(Maybe we should rename `ProtoBlock` to `DeployList` as part of this change?)

Reply:
We should know the parent block at the moment of proposal, shouldn't we? I mean, it should at least be possible to get the parent block hash from somewhere. We could then make the distinction between different blocks with the same `ProtoBlock`s by parent hash.

Good point, though.

And I have nothing against the renaming :+1:

Reply:
It would certainly be possible to put a parent hash into the `DeployList`; however, it would be kind of redundant and I think it would be far from trivial, given how the code is currently organized.

We currently don't have the concept of a proto block/deploy list parent. Introducing it would mean to keep track of almost a "parallel" preliminary chain of deploy lists. Since we currently need to be able to recreate a deploy list from a linear chain block, we'd end up putting that second "deploy list parent hash" into linear chain blocks, in addition to the proper parent _block_ hash…

Anyway, I don't think we should do that.

Reply:
I don't mean putting the parent hash into the `DeployList`. Rather, when we notify the Block Proposer about a proposal or finalization, we supply the parent block hash _along_ with the deploy list. Just so that we know that _this_ empty deploy list is a _different_ deploy list from _that_ empty deploy list. Or something along these lines.

Reply:
But if it's three subsequent empty blocks, we have two identical pairs of deploy lists.
I guess we could also use the block height or something. Nor sure.

Reply:
I am not sure about the architecture, so please forgive me if the following question is naive.

Can I spam finalized blocks (perhaps dated in the future)?  What would stop me from flooding the proposed `finalization_queue`?

Reply:
> I don't mean putting the parent hash into the `DeployList`. Rather, when we notify the Block Proposer about a proposal or finalization, we supply the parent block hash along with the deploy list. Just so that we know that this empty deploy list is a different deploy list from that empty deploy list.

Also: if I notified the Block Proposer with `(parent_block_hash, deploylist1)` and `(parent_block_hash, deploylist2)`, would it break anything?

(thank you again for your patience with my dumb questions)

Reply:
Chatting with @afck , we can't spam finalized blocks because they have 2/3rds of the validator signatures